apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfSubscriber.cs

```diff
@@ -0,0 +1,723 @@
+
+            private SubscriptionName _subscriptionName;
+            private SubscriberClient _client;
+            private readonly Func<PubsubMessage, CancellationToken, Task<Reply>> _handlerAsync;
+            private readonly TimeSpan _autoExtendInterval;
+            private readonly int _modifyDeadlineSeconds;
+            private readonly int _maxAckExtendCount;
+            private readonly Flow _flow;
+            private readonly Action<Task> _registerTaskFn;
+            private readonly IScheduler _scheduler;
+            private readonly TaskHelper _taskHelper;
+            private readonly CancellationTokenSource _softStopCts;
+            private readonly CancellationTokenSource _hardStopCts;
+
+            private readonly object _qLock = new object();
+            private readonly Queue<string> _ackQueue = new Queue<string>();
+            private readonly Queue<string> _extendQueue = new Queue<string>();
+            private int _unAckedMsgCount;
+            private TaskCompletionSource<int> _qEvent = new TaskCompletionSource<int>();
+
+            // States:
+            // Push faulted: local immediate stop/restart streaming-pull request
+            // - push task faults, which cancels streamingPullCts
+            // Pull faulted: local immediate stop/restart streaming-pull request
+            // - pull task faults, which cancels streamingPullCts
+            // soft-stop requested: no more msg processing/receiving, stop pull; continue push, then local hard-stop when all acks done
+            // - Pull task completes, which does not cancel streamingPullCts
+            // hard-stop requested: immediate stop of pull and push
+            // - Pull and Push fault, if they are still running
+
+            public async Task StartAsync()
+            {
+                // Loop to restart pull if it fails recoverably.
+                while (!_hardStopCts.IsCancellationRequested)
+                {
+                    SubscriberClient.StreamingPullStream pull = null;
+                    try
+                    {
+                        var streamingPullCts = CancellationTokenSource.CreateLinkedTokenSource(_hardStopCts.Token);
+                        pull = _client.StreamingPull(CallSettings.FromCancellationToken(streamingPullCts.Token),
+                            new BidirectionalStreamingSettings(10));// TODO: Put back to 1 once bug #188 released in v2.1.0
+                        // Initial call to start subscribe messages arriving.
+                        await _taskHelper.ConfigureAwait(pull.WriteAsync(new StreamingPullRequest
+                        {
+                            SubscriptionAsSubscriptionName = _subscriptionName,
+                            StreamAckDeadlineSeconds = _modifyDeadlineSeconds
+                        }));
+                        // Start Push task which pushes acks and extends.
+                        var pushTask = _taskHelper.Run(() => Push(pull, streamingPullCts.Token));
```

Comment:
Maybe streamingPullCts should be renamed given that it's used for pushing as well?